internal/resources/providers/gcplib/inventory/provider.go

```diff
@@ -466,35 +463,42 @@ func extendWithECS(ctx context.Context, crm *ResourceManagerWrapper, cache map[s
 }
 
 func (p *Provider) ListProjectsAncestorsPolicies(ctx context.Context) ([]*ProjectPoliciesAsset, error) {
-	projects := getAllAssets(p.log, p.inventory.ListAssets(ctx, &assetpb.ListAssetsRequest{
+	projects := p.getAllAssets(ctx, &assetpb.ListAssetsRequest{
 		ContentType: assetpb.ContentType_IAM_POLICY,
 		Parent:      p.config.Parent,
 		AssetTypes:  []string{CrmProjectAssetType},
-	}))
-
+	})
+	p.log.Infof("Listed %d GCP projects", len(projects))
+	ancestorsPolicies := map[string][]*ExtendedGcpAsset{}
 	return lo.Map(projects, func(project *assetpb.Asset, _ int) *ProjectPoliciesAsset {
 		projectAsset := extendWithECS(ctx, p.crm, p.crmCache, []*assetpb.Asset{project})[0]
 		// Skip first ancestor it as we already got it
-		policiesAssets := append([]*ExtendedGcpAsset{projectAsset}, getAncestorsAssets(ctx, p, project.Ancestors[1:])...)
+		policiesAssets := append([]*ExtendedGcpAsset{projectAsset}, getAncestorsAssets(ctx, ancestorsPolicies, p, project.Ancestors[1:])...)
 		return &ProjectPoliciesAsset{CloudAccount: projectAsset.CloudAccount, Policies: policiesAssets}
 	}), nil
 }
 
-func getAncestorsAssets(ctx context.Context, p *Provider, ancestors []string) []*ExtendedGcpAsset {
+func getAncestorsAssets(ctx context.Context, ancestorsPolicies map[string][]*ExtendedGcpAsset, p *Provider, ancestors []string) []*ExtendedGcpAsset {
 	return lo.Flatten(lo.Map(ancestors, func(parent string, _ int) []*ExtendedGcpAsset {
+		if ancestorsPolicies[parent] != nil {
+			return ancestorsPolicies[parent]
+		}
 		var assetType string
 		if strings.HasPrefix(parent, "folders") {
 			assetType = CrmFolderAssetType
 		}
 		if strings.HasPrefix(parent, "organizations") {
 			assetType = CrmOrgAssetType
 		}
-		assets := getAllAssets(p.log, p.inventory.ListAssets(ctx, &assetpb.ListAssetsRequest{
+
+		assets := p.getAllAssets(ctx, &assetpb.ListAssetsRequest{
 			ContentType: assetpb.ContentType_IAM_POLICY,
 			Parent:      parent,
 			AssetTypes:  []string{assetType},
-		}))
-		return extendWithECS(ctx, p.crm, p.crmCache, assets)
+		})
+		extendedAssets := extendWithECS(ctx, p.crm, p.crmCache, assets)
```

Comment:
Is there an option of reaching a rate limit with the CRM client as well?
for example, if the customer has many different projects...

If this is an option, we might need to add a limiter or at least a retrier.

Reply:
it looks like the `Get` operation which we use doesn't have a [quota](https://cloud.google.com/resource-manager/docs/limits). the only quotas i see are for `list` and `search`. (v3)

there is however room for improvements here, as `extendWithECS` uses a cache, but the underlining calls to get display names do not. so if for example, `projects/123/orgs/456` gets a cache hit in `extendWithECS` but `projects/789/orgs/456` does not, we still make a call to get the org display name, even though we already have it. 

i'll add caching for display names fetching as well. 

regarding retry - the `Project.Get(..).Context(..).Do()` call doesn't (AFAIK) enable us to retry (`Do()` only gets `key` and `value` back). we could replace the http client the CRM service uses, and account for retry there, but given there doesn't seem to be a quota for our usage (`Get()` only) i think maybe a cache would be sufficient. wdyt?

Reply:
yeah, If there is no quota, a thin cache layer will be sufficient